examples/mat-transpose.cpp

```diff
@@ -0,0 +1,314 @@
+      //
+      for (int ty = 0; ty < inner_Dim1; ++ty) {
+        for (int tx = 0; tx < inner_Dim0; ++tx) {
+
+          int col = bx * TILE_DIM + tx;  // Matrix column index
+          int row = by * TILE_DIM + ty;  // Matrix row index
+          if(row < N_rows and col < N_cols){
+            TILE[ty][tx] = Aview(row, col);
+          }
+          //std::cout<<Aview(row,col)<<std::endl;
+        }
+      }
+      //
+      // (2) Inner loops to read data from the tile
+      //
+      for (int ty = 0; ty < inner_Dim1; ++ty) {
+        for (int tx = 0; tx < inner_Dim0; ++tx) {
+
+          int col = by * TILE_DIM + tx;  // Transposed matrix column index
+          int row = bx * TILE_DIM + ty;  // Transposed matrix row index
+          if(row < N_cols and col < N_rows){
+            Atview(row, col) = TILE[tx][ty];
+          }
+
+        }
+      }
+    }
+  }
+
+  checkResult<int>(Atview, N_rows, N_cols);
+  //printResult<int>(Atview, N_cols, N_rows);
+  //----------------------------------------------------------------------------//
+
+
+  //----------------------------------------------------------------------------//
+  //RAJA Shared memory version...
+  printf("Run RAJA shared memory version \n");
+
+  std::memset(At, 0, N_rows * N_cols * sizeof(int));
+
+  auto iSpace =
+    RAJA::make_tuple(RAJA::RangeSegment(0, inner_Dim0), RAJA::RangeSegment(0,inner_Dim1),
+                     RAJA::RangeSegment(0, outer_Dim0), RAJA::RangeSegment(0,outer_Dim1));
+
+
+  //Toy shared memory object - For proof of concept.
+  using SharedTile = RAJA::SharedMem<int, TILE_DIM, TILE_DIM>;
+  using mySharedMemory = RAJA::SharedMemWrapper<SharedTile>;
+  mySharedMemory myTile;
+  mySharedMemory myTile2;
+
+  using KERNEL_EXEC_POL =
+    RAJA::KernelPolicy<
+      RAJA::statement::For<3, RAJA::loop_exec,
+        RAJA::statement::For<2, RAJA::loop_exec,
+
+                             RAJA::statement::CreateShmem<
+
+         RAJA::statement::Collapse<RAJA::omp_parallel_collapse_exec,
+                                   RAJA::ArgList<0, 1>,
+                                   RAJA::statement::Lambda<0>
+                                   >,
+
+         RAJA::statement::Collapse<RAJA::omp_parallel_collapse_exec,
+                                   RAJA::ArgList<0, 1>,
+                                   RAJA::statement::Lambda<1>
+                                   >
+                               > //Close shared memory scope
+        >//for 2
+       >//for 3
+      >; //close policy list
+
+  RAJA::kernel_param<KERNEL_EXEC_POL>(iSpace,
+                                      RAJA::make_tuple(myTile, myTile2),
+
+      [=] (int tx, int ty, int bx, int by, mySharedMemory &myTile,  mySharedMemory &myTile2) {
+
+           int col = bx * TILE_DIM + tx;  // Matrix column index
+           int row = by * TILE_DIM + ty;  // Matrix row index
+          if(row < N_rows and col < N_cols){
+            (*myTile.SharedMem)(ty,tx)  = Aview(row, col);
```

Comment:
@artv3 How can we do this without the explicit tile/index calculation?

Reply:
We can merge the initialization step of CreateShmem into SetShmemWindow. I would have to re-work some of the internals in SetShmemWindow to have it work with the shared memory wrapper, but should be straight forward.